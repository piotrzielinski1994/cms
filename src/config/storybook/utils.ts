import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
