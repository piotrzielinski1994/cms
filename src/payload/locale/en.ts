const en = {
  common: {
    content: 'Content',
    layout: 'Layout',
    component: {
      singular: 'Component',
      plural: 'Components',
    },
    section: {
      singular: 'Section',
      plural: 'Sections',
    },
    related: {
      singular: 'Related',
      plural: 'Related',
    },
    openInNewTab: 'Open in new tab',
    item: {
      singular: 'Item',
      plural: 'Items',
    },
    files: {
      singular: 'File',
      plural: 'Files',
    },
    default: 'Default',
    dark: 'Dark',
    question: 'Question',
    answer: 'Answer',
    pageNotFound: 'This page could not be found.',
    goHome: 'Go home',
  },
  fields: {
    label: 'Label',
    slug: 'Slug',
    path: 'Path',
    type: 'Type',
    title: 'Title',
    heading: 'Heading',
    subheading: 'Subheading',
    description: 'Description',
    image: 'Image',
    publishedAt: 'Published at',
    fullName: 'Full name',
    parent: 'Parent element',
    children: 'Child elements',
    role: {
      singular: 'Role',
      plural: 'Roles',
    },
    menu: 'Menu',
    submenu: 'Submenu',
    navItem: 'Navigation item',
    internalLink: 'Internal link',
    documentToLinkTo: 'Document to link to',
    buttons: 'Buttons',
    selector: 'Selector',
    isReversed: 'Reversed layout',
    subpages: 'Subpages',
    seo: 'SEO',
    acceptLabel: 'Accept label',
    readMoreLabel: 'Read more label',
    url: 'URL',
  },
  groups: {
    admin: 'Admin',
    docs: 'Docs',
  },
  collections: {
    pages: {
      singular: 'Page',
      plural: 'Pages',
    },
    categories: {
      singular: 'Category',
      plural: 'Categories',
    },
    posts: {
      singular: 'Post',
      plural: 'Posts',
    },
    sites: {
      singular: 'Site',
      plural: 'Sites',
    },
    users: {
      singular: 'User',
      plural: 'Users',
    },
    images: {
      singular: 'Image',
      plural: 'Images',
    },
    redirects: {
      singular: 'Redirect',
      plural: 'Redirects',
    },
  },
  enums: {
    role: {
      user: 'User',
      editor: 'Editor',
      admin: 'Admin',
    },
  },
  components: {
    header: 'Header',
    footer: 'Footer',
    cookiesBanner: 'Cookies banner',
  },
  frontend: {
    submit: 'Submit',
    logo: 'Logo',
    localeSwitcher: 'Locale switcher',
    themeSwitcher: 'Theme switcher',
    fontScaleSwitcher: 'Font scale switcher',
    increment: 'Increment',
    decrement: 'Decrement',
    close: 'Close',
    component: {
      skipLink: 'Go to the main content',
    },
    contactForm: {
      fields: {
        email: {
          label: 'Email',
        },
        message: {
          label: 'Message',
        },
      },
    },
  },
  zod: {
    invalid_type: 'Expected {expected}, received {received}',
    invalid_type_with_path: '{path} is expected {expected}, but received {received}',
    invalid_type_received_undefined: 'Required',
    invalid_literal: 'Invalid literal value, expected {expected}',
    unrecognized_keys: 'Unrecognized key(s) in object: {- keys}',
    unrecognized_keys_one: 'Unrecognized key in object: {- keys}',
    unrecognized_keys_other: 'Unrecognized keys in object: {- keys}',
    invalid_union: 'Invalid input',
    invalid_union_discriminator: 'Invalid discriminator value. Expected {- options}',
    invalid_enum_value: 'Invalid enum value. Expected {- options}, received {received}',
    invalid_arguments: 'Invalid function arguments',
    invalid_return_type: 'Invalid function return type',
    invalid_date: 'Invalid date',
    custom: 'Invalid input',
    invalid_intersection_types: 'Intersection results could not be merged',
    not_multiple_of: 'Number must be a multiple of {multipleOf}',
    not_finite: 'Number must be finite',
    invalid_string: {
      email: 'Invalid {validation}',
      url: 'Invalid {validation}',
      uuid: 'Invalid {validation}',
      cuid: 'Invalid {validation}',
      regex: 'Invalid',
      datetime: 'Invalid {validation}',
      startsWith: 'Invalid input: must start with {startsWith}',
      endsWith: 'Invalid input: must end with {endsWith}',
    },
    too_small: {
      array: {
        inclusive: 'Array must contain at least {minimum} element(s)',
        inclusive_one: 'Array must contain at least {minimum} element',
        inclusive_other: 'Array must contain at least {minimum} elements',
        not_inclusive: 'Array must contain more than {minimum} element(s)',
        not_inclusive_one: 'Array must contain more than {minimum} element',
        not_inclusive_other: 'Array must contain more than {minimum} elements',
      },
      string: {
        inclusive: 'String must contain at least {minimum} character(s)',
        inclusive_one: 'String must contain at least {minimum} character',
        inclusive_other: 'String must contain at least {minimum} characters',
        inclusive_with_path: '{path} must contain at least {minimum} character(s)',
        inclusive_with_path_one: '{path} must contain at least {minimum} character',
        inclusive_with_path_other: '{path} must contain at least {minimum} characters',
        not_inclusive: 'String must contain over {minimum} character(s)',
        not_inclusive_one: 'String must contain over {minimum} character',
        not_inclusive_other: 'String must contain over {minimum} characters',
        not_inclusive_with_path: '{path} must contain over {minimum} character(s)',
        not_inclusive_with_path_one: '{path} must contain over {minimum} character',
        not_inclusive_with_path_other: '{path} must contain over {minimum} characters',
      },
      number: {
        inclusive: 'Number must be greater than or equal to {minimum}',
        inclusive_with_path: '{path} must be greater than or equal to {minimum}',
        not_inclusive: 'Number must be greater than {minimum}',
        not_inclusive_with_path: '{path} must be greater than {minimum}',
      },
      set: {
        inclusive: 'Invalid input',
        not_inclusive: 'Invalid input',
      },
      date: {
        inclusive: 'Date must be greater than or equal to {- minimum, datetime}',
        not_inclusive: 'Date must be greater than {- minimum, datetime}',
      },
    },
    too_big: {
      array: {
        inclusive: 'Array must contain at most {maximum} element(s)',
        inclusive_one: 'Array must contain at most {maximum} element',
        inclusive_other: 'Array must contain at most {maximum} elements',
        not_inclusive: 'Array must contain less than {maximum} element(s)',
        not_inclusive_one: 'Array must contain less than {maximum} element',
        not_inclusive_other: 'Array must contain less than {maximum} elements',
      },
      string: {
        inclusive: 'String must contain at most {maximum} character(s)',
        inclusive_one: 'String must contain at most {maximum} character',
        inclusive_other: 'String must contain at most {maximum} characters',
        inclusive_with_path: '{path} must contain at most {maximum} character(s)',
        inclusive_with_path_one: '{path} must contain at most {maximum} character',
        inclusive_with_path_other: '{path} must contain at most {maximum} characters',
        not_inclusive: 'String must contain under {maximum} character(s)',
        not_inclusive_one: 'String must contain under {maximum} character',
        not_inclusive_other: 'String must contain under {maximum} characters',
        not_inclusive_with_path: '{path} must contain under {maximum} character(s)',
        not_inclusive_with_path_one: '{path} must contain under {maximum} character',
        not_inclusive_with_path_other: '{path} must contain under {maximum} characters',
      },
      number: {
        inclusive: 'Number must be less than or equal to {maximum}',
        inclusive_with_path: '{path} must be less than or equal to {maximum}',
        not_inclusive: 'Number must be less than {maximum}',
        not_inclusive_with_path: '{path} must be less than {maximum}',
      },
      set: {
        inclusive: 'Invalid input',
        not_inclusive: 'Invalid input',
      },
      date: {
        inclusive: 'Date must be smaller than or equal to {- maximum, datetime}',
        not_inclusive: 'Date must be smaller than {- maximum, datetime}',
      },
    },
  },
  storybook: {
    basic: {
      form: {
        textInput: {
          default: {
            label: 'Text Input',
            placeholder: 'Type value',
          },
          disabled: {
            label: 'Disabled Input',
            placeholder: 'Disabled',
          },
          invalid: {
            label: 'Invalid Input',
            value: 'Invalid value',
            error: 'Error message',
          },
        },
        numberInput: 'Number Input',
        textArea: 'Text Area',
        checkbox: {
          default: {
            label: 'Checkbox',
          },
          disabled: {
            label: 'Disabled Checkbox',
          },
          invalid: {
            label: 'Invalid Checkbox',
            error: 'Error message',
          },
        },
        radio: {
          default: {
            label: 'Radio Button',
          },
          disabled: {
            label: 'Disabled Radio Button',
          },
          invalid: {
            label: 'Invalid Radio Button',
            error: 'Error message',
          },
        },
        select: {
          label: 'Select',
          placeholder: 'Choose value',
          option: 'Option',
        },
      },
      button: {
        default: 'Button',
        disabled: 'Disabled',
        secondary: 'Secondary',
      },
      alert: {
        info: 'Information',
        success: 'Success',
        warn: 'Warning',
        error: 'Error',
      },
      badge: 'Badge',
      accordion: {
        heading: 'Heading',
        content: 'Content',
      },
      dialog: {
        button: 'Show Dialog',
        header: 'Header',
        content: 'Content',
        submit: 'Submit',
        cancel: 'Cancel',
      },
    },
  },
};

export { en };
