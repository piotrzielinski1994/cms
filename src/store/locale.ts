import { contentLocales, defaultContentLocale } from '@/config/locales.config';
import { Locale, useLocale } from 'next-intl';

const useLocaleStore = (): Locale => {
  const locale = useLocale();
  return contentLocales.includes(locale) ? locale : defaultContentLocale;
};

export { useLocaleStore };
