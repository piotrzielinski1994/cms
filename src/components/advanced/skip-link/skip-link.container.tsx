import { Optional } from '@/utils/types';
import { useTranslations } from 'next-intl';
import { ComponentProps } from 'react';
import { SkipLink } from './skip-link';

const SkipLinkContainer = (props: Optional<ComponentProps<typeof SkipLink>, 't'>) => {
  const t = useTranslations('frontend');
  return <SkipLink {...props} t={{ label: t('component.skipLink'), ...props.t }} />;
};

export { SkipLinkContainer };
