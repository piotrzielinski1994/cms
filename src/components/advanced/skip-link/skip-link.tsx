'use client';

import { ButtonLink } from '@/components/basic/button/button';
import { cn } from '@/utils/tailwind';
import { useTranslations } from 'next-intl';

const SkipLink = () => {
  const t = useTranslations('frontend');
  return (
    <ButtonLink
      href="#main"
      aria-hidden={true}
      onFocus={(e) => e.currentTarget.setAttribute('aria-hidden', 'false')}
      onBlur={(e) => e.currentTarget.setAttribute('aria-hidden', 'true')}
      className={cn(
        'fixed left-4 z-skipLink',
        '-translate-y-full focus:translate-y-4 transition-transform',
        'w-fit',
      )}
    >
      {t('component.skipLink')}
    </ButtonLink>
  );
};

export { SkipLink };
