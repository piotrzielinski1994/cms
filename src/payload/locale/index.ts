import { clientEnv } from '@/env.client';
import {
  DefaultTranslationKeys,
  NestedKeysStripped,
  SupportedLanguages,
  TFunction,
} from '@payloadcms/translations';
import { en } from '@payloadcms/translations/languages/en';
import { pl } from '@payloadcms/translations/languages/pl';
import { pick } from 'ramda';
import { en as customEn } from './en';
import { pl as customPl } from './pl';

// Types ====================================
export type AdminTranslations = TFunction<
  NestedKeysStripped<typeof customEn> | DefaultTranslationKeys
>;

// Variables ====================================
export const adminLocale = {
  list: pick(clientEnv.feature.locale.admin.list, { en, pl } as SupportedLanguages),
  customList: {
    en: customEn,
    pl: customPl,
  },
  default: clientEnv.feature.locale.admin.default,
};

export const contentLocale = {
  list: clientEnv.feature.locale.content.list,
  default: clientEnv.feature.locale.content.default,
};
