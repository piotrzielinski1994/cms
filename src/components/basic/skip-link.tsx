import { useTranslationsStore } from '@/store/translations';
import { cn } from '@/utils/tailwind';
import { ButtonLink } from './button';

const SkipLink = () => {
  const t = useTranslationsStore();
  return (
    <ButtonLink
      href="#main"
      className={cn(
        'fixed left-4 z-skipLink',
        '-translate-y-full focus:translate-y-4 transition-transform',
        'w-fit',
      )}
    >
      {t.component.skipLink}
    </ButtonLink>
  );
};

export { SkipLink };
