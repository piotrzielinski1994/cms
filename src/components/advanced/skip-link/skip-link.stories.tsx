import Form from '@/components/basic/form/root/form';
import { TextInput } from '@/components/basic/form/text-input/text-input';
import { Link } from '@/components/basic/link/link';
import { cn } from '@/utils/tailwind';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useEffect, useRef, type ComponentProps } from 'react';
import { SkipLink as SkipLinkComponent } from './skip-link';

type Args = ComponentProps<typeof SkipLinkComponent> & {
  label?: string;
};

const meta: Meta<Args> = {
  component: SkipLinkComponent,
  title: 'Advanced/SkipLink',
};

const Render = () => {
  const t = useTranslations('storybook.advanced.skipLink');
  const t2 = useTranslations('frontend');
  const skipLinkRef = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    // Programmatically prevent default Storybook's behavior after clicking link
    const el = skipLinkRef.current;
    if (!el) return;

    const focusMain = (e: MouseEvent) => {
      e.preventDefault();
      document.getElementById('main')?.focus();
    };

    el.addEventListener('click', focusMain);
    return () => el.removeEventListener('click', focusMain);
  }, []);

  return (
    <>
      {/* Compensate storybook's padding on body */}
      <div className="-mt-4">
        <SkipLinkComponent ref={skipLinkRef} t={{ label: t2('component.skipLink') }} />
      </div>

      <nav className={cn('flex justify-between gap-x-4 flex-wrap', 'p-2 bg-background1')}>
        <span>{t('navbar')}</span>
        <ul className="flex justify-end gap-2">
          {Array.from({ length: 3 }).map((_, index) => {
            return (
              <li key={index}>
                <Link href="">{`${t('link')} ${index + 1}`}</Link>
              </li>
            );
          })}
        </ul>
      </nav>

      <main className="focus-visible:tw-cms-outline" id="main" tabIndex={-1}>
        <Form.Group>
          <Form.Label htmlFor="input">{t('main')}</Form.Label>
          <TextInput id="input" name="input" placeholder={t('inputPlaceholder')} />
        </Form.Group>
      </main>
    </>
  );
};

const SkipLink: StoryObj<typeof SkipLinkComponent> = { render: Render };

export { SkipLink };
export default meta;
