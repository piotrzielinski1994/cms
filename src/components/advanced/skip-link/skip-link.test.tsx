import { en } from '@/payload/locale/en';
import { withProviders } from '@/utils/tests';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { SkipLink } from './skip-link';

const t = en.frontend.component;

describe('SkipLink', () => {
  it('should have no accessibility violations', async () => {
    const { container } = render(withProviders(<SkipLink />));
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(withProviders(<SkipLink />));
    expect(container).toMatchSnapshot();
  });

  describe('Interaction', () => {
    beforeEach(() => {
      window.history.pushState({}, '', ' ');
    });

    it('should be hidden if not focused', () => {
      const { getByRole } = render(withProviders(<SkipLink />));
      const link = getByRole('link', { hidden: true });
      expect(link).toHaveTextContent(t.skipLink);
    });

    it('should be visible if focused', async () => {
      const { getByRole } = render(withProviders(<SkipLink />));

      await userEvent.tab();

      const link = getByRole('link', { hidden: false });
      expect(link).toHaveTextContent(t.skipLink);
    });

    it('should focus #main when activated with Enter key', async () => {
      const { getByRole } = render(withProviders(<SkipLink />));
      const link = getByRole('link', { hidden: true });

      expect(window.location.hash).toBe('');

      await userEvent.tab();

      expect(link).toHaveFocus();

      await userEvent.keyboard('{Enter}');

      expect(window.location.hash).toBe('#main');
    });

    it('should not focus #main when activated with spacebar', async () => {
      const { getByRole } = render(withProviders(<SkipLink />));
      const link = getByRole('link', { hidden: true });

      expect(window.location.hash).toBe('');

      await userEvent.tab();

      expect(link).toHaveFocus();

      await userEvent.keyboard(' ');

      expect(window.location.hash).toBe('');
    });
  });
});
