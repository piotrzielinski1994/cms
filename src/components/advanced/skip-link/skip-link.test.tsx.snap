// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkipLink > should match the snapshot 1`] = `
<div>
  <a
    aria-hidden="true"
    class="px-6 py-3 border-2 border-primary hover:border-primary/90 font-semibold whitespace-nowrap disabled:cursor-not-allowed focus-visible:tw-cms-outline bg-primary hover:bg-primary/90 text-primary-foreground disabled:border-primary/0 disabled:bg-primary/50 fixed left-4 z-skipLink -translate-y-full focus:translate-y-4 transition-transform w-fit"
    href="#main"
  >
    Go to the main content
  </a>
</div>
`;
